docs/path-resolution.rst

```diff
@@ -0,0 +1,690 @@
+   :ref:`relative imports <relative-imports>` part of it comes from the source unit name of the
+   importing file.
+
+   .. index:: standard input, stdin, <stdin>
+
+#. **Standard input**
+
+   The last way to provide the source is by sending it to compiler's :ref:`standard input
+   <standard-input>`:
+
+   .. code-block:: bash
+
+       echo 'import "./util.sol"; contract C {}' | solc -
+
+   The content of the standard input is identified in the virtual filesystem by a special source unit name:
+   ``<stdin>``.
+
+.. warning::
+
+    The compiler uses source unit names to determine whether imports refer to the same source unit or not.
+    If you refer to a file in multiple ways that translate to different names, it will be compiled
+    multiple times.
+
+    For example:
+
+    .. code-block:: solidity
+        :caption: /code/contract.sol
+
+        import "tokens/token.sol" as token1;   // source unit name: tokens/token.sol
+        import "tokens///token.sol" as token2; // source unit name: tokens///token.sol
+
+    .. code-block:: bash
+
+        cd /code
+        solc contract.sol /code/tokens/token.sol # source unit name: /code/tokens/token.sol
+
+    In the above ``token.sol`` will end up in the virtual filesystem under three different
+    source unit names even though all the paths refer to the same file in the underlying filesystem.
+
+    To avoid this situation it is recommended to always use the canonical form of paths in your
+    imports and to only list the top-level files that are not imported by other files when
+    invoking the CLI compiler.
+
+.. index:: ! direct import, import; direct
+.. _direct-imports:
+
+Direct Imports
+==============
+
+An import that does not start with ``./`` or ``../`` is a *direct import*.
+
+::
+
+    import "/project/lib/util.sol";         // source unit name: /project/lib/util.sol
+    import "lib/util.sol";                  // source unit name: lib/util.sol
+    import "@openzeppelin/address.sol";     // source unit name: @openzeppelin/address.sol
+    import "https://example.com/token.sol"; // source unit name: https://example.com/token.sol
+
+The import path translates directly to a source unit name without normalization of any kind:
+
+::
+
+    import "/project/lib/../lib///math.sol"; // source unit name: /project/lib/../lib///math.sol
+    import "lib/../lib///math.sol";          // source unit name: lib/../lib///math.sol
+
+In the above you might expect the source unit names to be ``/project/lib/math.sol`` and
+``lib/math.sol`` respectively but this is not the case.
+For direct imports the source unit name is exactly what is stated in the import.
+When the source is provided via Standard JSON interface each of these names can actually be
+associated with different content.
+
+When the source is not available in the virtual filesystem, the compiler passes the source unit name
+to the file loader.
+The default loader will attempt to use it as a path and look up the file on disk.
```

Comment:
is this true for all binaries?

Reply:
Well, technically yes because the default loader is present in all binaries but some binaries just do not use it :)

Anyway, this is now reworded after I added a section on the callback.